src/ast/irbuilderbpf.cpp

```diff
@@ -362,15 +362,30 @@ Value *IRBuilderBPF::CreateUSDTReadArgument(Value *ctx, struct bcc_usdt_argument
       abort();
     }
 
+    // bpftrace's args are internally represented as 64 bit integers. However,
+    // the underlying argument (of the target program) may be less than 64
+    // bits. So we must be careful to zero out unused bits.
     Value* reg = CreateGEP(ctx, getInt64(offset * sizeof(uintptr_t)), "load_register");
     AllocaInst *dst = CreateAllocaBPF(builtin.type, builtin.ident);
-    CreateProbeRead(dst, builtin.type.size, reg);
-    result = CreateLoad(dst);
     if (argument->valid & BCC_USDT_ARGUMENT_DEREF_OFFSET) {
+      // If we are dereferencing a pointer for the offset, we want to
+      // read native pointer size (assumed to be 64-bit here)
+      assert(builtin.type.size == 8);
+      CreateProbeRead(dst, builtin.type.size, reg);
+      result = CreateLoad(dst);
       Value *ptr = CreateAdd(
           result,
           getInt64(argument->deref_offset));
-      CreateProbeRead(dst, builtin.type.size, ptr);
+      // Zero out `dst` here in case we read less than 64 bits
+      CreateStore(getInt64(0), dst);
+      CreateProbeRead(dst, argument->size, ptr);
```

Comment:
Signed types have [a negative value](https://sourceware.org/systemtap/wiki/UserSpaceProbeImplementation)

```suggestion
      CreateProbeRead(dst, std::abs(argument->size), ptr);
```

I think it would be better to check `argument->size` in case a binary contains an invalid value.